[FIX] account, l10n_it_edi: allow multiple invoices in the same file

lordkrandel opened the pull request:
Italian EDI FatturaPA XML files can be made of several invoice descriptions put together in the same file. We used to import just the first. Now we import all of them and show a list view of imported files.

This commit also changes how we check the file to be a proper FatturaPA, instead of looking at the filename, we now properly look at the XML structure, that now is being parsed just as we do for all other EDIs.

CADES signed files (.xml.p7m) have their signature removed and then they are treated like normal XML files.

In order to be able to do this, we have to move the code that specially handles the account_predictive_bills for Italy to the `account` level.

Related PR: odoo/odoo#163849
Related issue: https://github.com/odoo/odoo/issues/150382
Task [link](https://www.odoo.com/web#model=project.task&id=3972965)
task-3972965

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/165673).

lordkrandel commented:
Opened a draft for discussion.
The behaviour of the module is non the same as other EDIs at the moment.

lordkrandel commented on addons/account/models/ir_attachment.py:
```diff
@@ -34,10 +34,12 @@ def _decode_edi_xml(self, filename, content):
         :returns:           A list with a dictionary.
         """
+
+        parser = etree.XMLParser(resolve_entities=False)
```
This is a work in progress.

This was a problem for Italian clients, as the XMLs are also imported on third party modules' tests that have no internet connections, and the entities sometimes are not local and require the Internet to be resolved.

lordkrandel commented on addons/account/models/ir_attachment.py:
```diff
@@ -34,10 +34,12 @@ def _decode_edi_xml(self, filename, content):
+            xml_tree = etree.fromstring(content, parser)
         except Exception as e:
-            _logger.exception("Error when converting the xml content to etree: %s", e)
+            _logger.warning("Error when converting the xml content to etree: %s", e, exc_info=True)
```
Why `_logger.exception` while the binary decoder does `_logger.warning` ?

lordkrandel commented:
Think I found a bug.

Italian XML can have multiple invoices per XML file.
See the `FatturaElettronicaBody` tag
![image](https://github.com/odoo/odoo/assets/1665365/554bb147-47fe-4ecb-b215-76257810d94a)

At the moment, I think that this code here will block the double import, because we have the same attachment for two `file_data` structures.
https://github.com/odoo/odoo/blob/master/addons/account/models/account_move.py#L3498

I also don't know if we're able to show the two invoices, should we even be able to import them.

lordkrandel commented on addons/account/models/account_journal.py:
```diff
@@ -866,22 +866,28 @@ def _create_document_from_attachment(self, attachment_ids):
         # As we are coming from the journal, we assume that each attachments
         # will create an invoice with a tentative to enhance with EDI / OCR..
         all_invoices = self.env['account.move']
+        Move = self.env['account.move'].with_company(self.company_id)
```
You're right.
*It was a workaround for another problem, but now I solved that so this is useless*

lordkrandel replied:
No... actually the second move geets created with `company_1_data` even if the journal is on `company_2_data` because the `get_edi_creation` is called on `current_invoice` which takes the `self.env.company`'s  defaults. 
So what I did here was actually necessary :| ...? Halp?

![image](https://github.com/odoo/odoo/assets/1665365/02e08631-81a4-4bb7-8cd1-2ed750735cb7)

lordkrandel replied:
Done.

smetl commented:
@lordkrandel Can you rebase plz? :)

lordkrandel commented:
@tsb-odoo  ping.
Testable with [this file](https://github.com/odoo/odoo/pull/165673/files#diff-a17f99cd0e5ef5c0970d114e7bb82a14eefc9098827f68a6e8db9175316d31f3) if you change the company's `vat` to `02780790107` (sam as the one in the test file)

lordkrandel commented:
@smetl  TSB marked the task as green, so I moved this to review.
I have updated the commit message and the description, as this is mainly a bugfix than an improvement.

lordkrandel commented:
ping @odoo/rd-security

mart-e commented:
ping @lordkrandel

lordkrandel commented:
👋🏻

william-andre commented:
@lordkrandel I think that the process of the security would be faster if you could provide for each warning why you believe that there is no actual issue :wink:

lordkrandel commented:
I think the PR is ready for @odoo/rd-security review.

If you want more detail about what to look for,

> [addons/account/models/ir_attachment.py:39](https://github.com/odoo/odoo/blob/3351a11ffa279df3f5cdfdf143739b0a4c4a6beb/addons/account/models/ir_attachment.py#L39-L39)
[PY012 🔗](https://github.com/odoo/enterprise/wiki/security#PY012): parser = etree.XMLParser(resolve_entities=False)

- `XMLParser` has been added the `resolve_entities=False`, actually fixing `PY012` as it requires it by definition.

> [addons/l10n_it_edi/models/ir_attachment.py:13](https://github.com/odoo/odoo/blob/3351a11ffa279df3f5cdfdf143739b0a4c4a6beb/addons/l10n_it_edi/models/ir_attachment.py#L13-L13)
[PY004 🔗](https://github.com/odoo/enterprise/wiki/security#PY004): for suffix in Path(filename).suffixes[::-1]:
> [addons/l10n_it_edi/models/ir_attachment.py:72](https://github.com/odoo/odoo/blob/3351a11ffa279df3f5cdfdf143739b0a4c4a6beb/addons/l10n_it_edi/models/ir_attachment.py#L72-L72)
[PY004 🔗](https://github.com/odoo/enterprise/wiki/security#PY004): suffixes = [suffix.lower() for suffix in Path(filename).suffixes]

- Added a method to emulate `os.path.basename` with `Path.suffixes[::-1]`. I dislike importing os.path, and it would anyway would have been a security issue anyway. I just need a `file.xml.whatever.p7m` to which we remove the signature, to be stored as an attachment with a simple `file.xml` filename.

I don't know what else to add, let me know if you have doubts. I just find my explanation more unclear than the code itself, but whatever, if it helps you, good for you.

mart-e commented on addons/account/models/account_move.py:
```diff
@@ -3127,6 +3127,11 @@ def _get_edi_decoder(self, file_data, new=False):
             return lambda *args: False
         return
 
+    @api.model
+    def _is_prediction_enabled(self):
+        """ The account_accountant module overwrites new lines with predicted products, taxes and accounts. """
+        return 'payment_state_before_switch' in self._fields
+
```
Could be a flag set to `False` and changed to `True` in `account_accountant`?

lordkrandel replied:
Don't remember, I removed the function and inlined the code in `_l10n_it_edi_import_line`, if there's a problem, tests should tell.

william-andre commented on addons/account/models/account_move.py:
```diff
@@ -3199,25 +3208,26 @@ def add_file_data_results(file_data, invoice):
+                for file_data_part in get_parts(file_data):
+                    try:
+                        with self.env.cr.savepoint():
+                            invoice = current_invoice or self.with_company(company).create({})
```
> as more than 64 savepoints in a transaction will spill the entire thing

Can even be cross transaction actually, and when/how the locks are released AFAIK

lordkrandel commented:
> > * `XMLParser` has been added the `resolve_entities=False`, actually fixing `PY012` as it requires it by definition.
> 
> It's completely useless, just use `etree.fromstring` without a custom parser and everything works fine.

It doesn't work fine. This is from the IAP but it's the same file-reading problem.
![immagine](https://github.com/user-attachments/assets/2e53dacc-0852-431e-a9d6-c99cc42b010c)
This is on the IAP but still.

lordkrandel commented:
> > * `XMLParser` has been added the `resolve_entities=False`, actually fixing `PY012` as it requires it by definition.
> 
> It's completely useless, just use `etree.fromstring` without a custom parser and everything works fine.

I had a fast reply, but thinking again, I have to verify. 
I had a case in the IAP but to be sincere I don't remember if a Parser was given or not.

lordkrandel commented:
> > * `XMLParser` has been added the `resolve_entities=False`, actually fixing `PY012` as it requires it by definition.
> 
> It's completely useless, just use `etree.fromstring` without a custom parser and everything works fine.

@xmo-odoo , I've done some checks, looks like it's no good in `lxml < 5`, which is what we have for up to `Python 3.11` in `requirements.txt`

I probably also should add `recover=True` since we have lots of occurrences of the last image's error on the IAP where we also parse invoices (and then port it on the IAP)


![immagine](https://github.com/user-attachments/assets/2ea01e05-d4bc-4bc0-86d5-47d6954f6e79)

![immagine](https://github.com/user-attachments/assets/965401fc-bc47-4b63-88f5-c78bfc8c8f1d)

![immagine](https://github.com/user-attachments/assets/be586bb8-65cf-46f1-a8b7-62d1d7142b0b)

![immagine](https://github.com/user-attachments/assets/8990436c-6293-42e2-966b-744c8f67076d)

lordkrandel commented:
> > > > * `XMLParser` has been added the `resolve_entities=False`, actually fixing `PY012` as it requires it by definition.
> > > 
> > > 
> > > It's completely useless, just use `etree.fromstring` without a custom parser and everything works fine.
> > 
> > 
> > I've done some checks, looks like it's no good in `lxml < 5`, which is what we have for up to `Python 3.11` in `requirements.txt`
> 
> https://github.com/odoo/odoo/blob/39d9a29cd6f8e0b334a0c8e788edc5340038d2c5/odoo/tools/misc.py#L118-L122

Ah! Ok, guns down, you win 😆
We might need the `recover` though, that issue is real, people have malformed URIs, though the XML is still considered valid by the Tax agency.

william-andre commented on addons/account/models/account_move.py:
```diff
@@ -3141,84 +3141,110 @@ def _extend_with_attachments(self, attachments, new=False):
+                    else:
+                        import_cr.rollback()
+
+        with closing(self.env.registry.cursor()) as import_cr:
```
It will only do so if/when it can. Other transactions using savepoints can also prevent cleaning up the memory
More details here
https://github.com/odoo/odoo/pull/71395#issuecomment-851440042

lordkrandel replied:
This PR is in a messy state.
If I have to doubleloop on parts of a file, then savepoint has to go, and it's not simple. To be true, I'd like to refactor the whole extend with attachments in master, since there's lot of code here just to filter out cases peculiar to one of the three callers (mail, post on invoice, import from journal) lots of continue etc. I asked Laurent some time but it look he s given this up to you Antoine.
The reason to refactor it is precisely to also have control on invoice creation.
In some cases we already have a preliminary function (i.e. create with attachments) so I don't really know why there is, for example, a "new" parameter here to spaghetti it up things a bit, not counting 5 continue s.
This is the core of invoice import, and the level of clarity and readabiliy is quite low.

lordkrandel replied:
Ideally, I'd like decoders to be generators that only extract dicts of values from the files, instead of manipulating invoices while reading. The creation or update can then just be a load_records, just like we do for chart templates.

The whole idea of creating a fixed number of invoices before reading the files is wrong imho, if you have errors (amounting to 0) or multiple invoices per file makes it difficult.

One of the things that make this passage  hard is that PDFs can contain XMLs that contain PDFs... and we might have to create the invoice from the XML, since it's the most structured data inside the file. If we do, then the rest of attachment (embedding included) must only be attached to the invoice, but no further be examined.

Another, is that if you have multiple attachments from the same mail, only the first that yields an invoice is considered, the rest is just attached.
But this doesn't really be a concern when you import from journal or attach to a message in the chatter

antoine162 replied:
I totally agree this would be a great thing to refactor in master (I mentioned it to Laurent and he agreed we need to refactor it - which is maybe why he pinged me? I'm not sure whether he wanted me to just have a look or whether to do a full review though.)

I also realize that removing the savepoint isn't simple. Comparing the various options, none seems perfectly satisfactory:
- savepoints: perf problem
- in the same transaction, commit / rollback after each import: it's not great because it can mess up the request retry mechanism in case of a serialization error, depending on what code was executed before calling the import code. In addition, it means we need some way of keeping track of what was already successfully imported. Or we need to manually catch serialization errors during import and do a rollback + retry in that case, which is messy.
- open new transactions like you suggested: similar problems, as we probably need to do a commit before opening the new transactions, and serialization errors can still occur so we need to handle them.
- delegate each import to a cron job: would solve the concurrency problems, but we don't have enough control over when the crons are executed. Plus, crons aren't designed to be called multiple times in different transactions (they are supposed to batch things as much as possible in a single transaction).

So, my last thought was, we can maybe create our own 'synchronous cron' just for this purpose? An object that allows you to schedule multiple jobs, each to be run in its own transaction, with retry mechanisms in case of serialization errors, and which would be run right when the request returns?

Maybe it would be a lot of development, but to me it seems like that would fit our needs.

antoine162 replied:
> Ideally, I'd like decoders to be generators that only extract dicts of values from the files, instead of manipulating invoices while reading. The creation or update can then just be a load_records, just like we do for chart templates.

That would be really nice, if it's feasible. I wonder though whether in practice we need to use the computes to progressively 'build up' an invoice. If we don't, that would definitely be the way to go, I think.

> The whole idea of creating a fixed number of invoices before reading the files is wrong imho, if you have errors (amounting to 0 invoices) or multiple invoices per file it makes things hard.
> 
> One of the things that make this passage hard is that PDFs can contain XMLs that contain PDFs... and we might have to create the invoice from the XML, since it's the most structured data inside the file. If we do, then the rest of attachment (embedding included) must only be attached to the invoice, but no further be examined.
> 
> Another, is that if you have multiple attachments from the same mail, only the first that yields an invoice is considered, the rest is just attached. But this doesn't really be a concern when you import from journal or attach to a message in the chatter

It's funny, I was thinking so too! I think a solution would be to more clearly separate the logic that determines the attachments to decode (including separating embedded XMLs) from the logic that calls the decoders on the right attachments.

lordkrandel commented:
Superseded by odoo/odoo#189979
